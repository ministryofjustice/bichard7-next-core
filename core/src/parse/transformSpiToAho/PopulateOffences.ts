import { lookupAlcoholLevelMethodBySpiCode, lookupResultQualifierCodeByCjsCode } from "../../dataLookup"
import { COMMON_LAWS, INDICTMENT } from "../../lib/offenceTypes"
import resultCodeIsOnStopList from "../../lib/result/resultCodeIsOnStopList"
import type { CriminalProsecutionReference, Offence, OffenceCode } from "../../types/AnnotatedHearingOutcome"
import type { OffenceParsedXml, ResultedCaseMessageParsedXml, SpiResult } from "../../types/SpiResult"
import PopulateOffenceResults from "./PopulateOffenceResults"
import removeSeconds from "./removeSeconds"

const enteredInErrorResultCode = 4583 // Hearing Removed
const dontKnowValue = "D"
const adjournmentSineDieResultCode = 2007
// TODO: Refactor into an enum
const timeRange = {
  ON_OR_IN: 1,
  BEFORE: 2,
  AFTER: 3,
  BETWEEN: 4,
  ON_OR_ABOUT: 5,
  ON_OR_BEFORE: 6
}

export interface OffencesResult {
  offences: Offence[]
  bailConditions: string[]
}

const adjournmentSineDieConditionMet = (spiResults: SpiResult[]) => {
  let a2007ResultFound = false
  let aFailConditionResultFound = false

  spiResults.forEach((result) => {
    const resultCode = result.ResultCode !== undefined ? Number(result.ResultCode) : undefined
    if (resultCode === adjournmentSineDieResultCode) {
      a2007ResultFound = true
    } else if (!resultCodeIsOnStopList(resultCode ?? 1000)) {
      aFailConditionResultFound = true
    }
  })

  return a2007ResultFound && !aFailConditionResultFound
}

export default class {
  private bailConditions: string[] = []

  constructor(
    private courtResult: ResultedCaseMessageParsedXml,
    private hearingDefendantBailConditions: string[] = [],
    private isAdjournmentSineDieConditionMet = false
  ) {}

  private getOffenceReason = (spiOffenceCode: string): OffenceCode => {
    const spiOffenceCodeLength = spiOffenceCode.length
    const offenceCode: Partial<OffenceCode> = {
      Reason: spiOffenceCodeLength > 4 ? spiOffenceCode.substring(4, Math.min(7, spiOffenceCodeLength)) : "",
      ...(spiOffenceCodeLength > 7 ? { Qualifier: spiOffenceCode.substring(7) } : {}),
      FullCode: spiOffenceCode
    }

    if (spiOffenceCode.startsWith(COMMON_LAWS)) {
      offenceCode.__type = "CommonLawOffenceCode"
      if (offenceCode.__type === "CommonLawOffenceCode") {
        offenceCode.CommonLawOffence = COMMON_LAWS
      }
    } else if (spiOffenceCode.startsWith(INDICTMENT)) {
      offenceCode.__type = "IndictmentOffenceCode"
      if (offenceCode.__type === "IndictmentOffenceCode") {
        offenceCode.Indictment = INDICTMENT
      }
    } else {
      offenceCode.__type = "NonMatchingOffenceCode"
      if (offenceCode.__type === "NonMatchingOffenceCode") {
        offenceCode.ActOrSource = spiOffenceCodeLength < 2 ? spiOffenceCode : spiOffenceCode.substring(0, 2)

        offenceCode.Year =
          spiOffenceCodeLength > 2 ? spiOffenceCode.substring(2, Math.min(4, spiOffenceCodeLength)) : ""
      }
    }

    return offenceCode as OffenceCode
  }

  private populateOffence = (spiOffence: OffenceParsedXml): Offence | undefined => {
    const offence = {} as Offence
    const {
      BaseOffenceDetails: {
        OffenceCode: spiOffenceCode,
        OffenceSequenceNumber: spiOffenceSequenceNumber,
        ArrestDate: spiArrestDate,
        ChargeDate: spiChargeDate,
        LocationOfOffence: spiLocationOfOffence,
        OffenceWording: spiOffenceWording,
        AlcoholRelatedOffence: spiAlcoholRelatedOffence,
        OffenceTiming: { OffenceDateCode: spiOffenceDateCode, OffenceStart: spiOffenceStart, OffenceEnd: spiOffenceEnd }
      },
      ConvictionDate: spiConvictionDate,
      Result: spiResults
    } = spiOffence

    const enteredInError = spiResults.some(
      (result) => result.ResultCode !== undefined && Number(result.ResultCode) === enteredInErrorResultCode
    )
    if (enteredInError) {
      return undefined
    }

    const OffenceCode = this.getOffenceReason(spiOffenceCode ?? "")
    offence.CriminalProsecutionReference = {
      OffenceReason: {
        OffenceCode,
        __type: "NationalOffenceReason"
      }
    } as CriminalProsecutionReference

    offence.ArrestDate = spiArrestDate ? new Date(spiArrestDate) : undefined
    offence.ChargeDate = spiChargeDate ? new Date(spiChargeDate) : undefined
    offence.ActualOffenceDateCode = spiOffenceDateCode?.toString()
    offence.ActualOffenceStartDate = {
      StartDate: new Date(spiOffenceStart.OffenceDateStartDate)
    }
    offence.ActualOffenceEndDate = spiOffenceEnd ? { EndDate: new Date(spiOffenceEnd.OffenceEndDate) } : undefined
    offence.LocationOfOffence = spiLocationOfOffence
    offence.ActualOffenceWording = spiOffenceWording

    if (spiAlcoholRelatedOffence) {
      offence.AlcoholLevel = {
        Amount: Number(spiAlcoholRelatedOffence.AlcoholLevelAmount),
        Method:
          lookupAlcoholLevelMethodBySpiCode(spiAlcoholRelatedOffence.AlcoholLevelMethod)?.cjsCode ??
          spiAlcoholRelatedOffence.AlcoholLevelMethod
      }
    }

    if (spiOffenceStart?.OffenceStartTime) {
      const spiOffenceStartTime = removeSeconds(spiOffenceStart.OffenceStartTime)

      const offenceDateCode = Number(spiOffenceDateCode)
      const { ON_OR_IN, BEFORE, AFTER, ON_OR_ABOUT, ON_OR_BEFORE, BETWEEN } = timeRange
      if ([ON_OR_IN, BEFORE, AFTER, ON_OR_ABOUT, ON_OR_BEFORE].includes(offenceDateCode)) {
        offence.OffenceTime = spiOffenceStartTime
      } else if (offenceDateCode === BETWEEN) {
        offence.StartTime = spiOffenceStartTime
      }
    }

    if (spiOffenceEnd?.OffenceEndTime) {
      offence.OffenceEndTime = removeSeconds(spiOffenceEnd.OffenceEndTime)
    }

    offence.CommittedOnBail = dontKnowValue
    offence.CourtOffenceSequenceNumber = Number(spiOffenceSequenceNumber)

    offence.ConvictionDate = spiConvictionDate ? new Date(spiConvictionDate) : undefined

    if (!spiConvictionDate) {
      this.isAdjournmentSineDieConditionMet ||= adjournmentSineDieConditionMet(spiResults)

      if (this.isAdjournmentSineDieConditionMet) {
        offence.ConvictionDate = new Date(this.courtResult.Session.CourtHearing.Hearing.DateOfHearing)
      }
    }

    const { results, bailQualifiers } = new PopulateOffenceResults(this.courtResult, spiOffence).execute()
    offence.Result = results

    if (this.hearingDefendantBailConditions.length > 0 && bailQualifiers.length > 0) {
      bailQualifiers.forEach((bailQualifier) => {
        const description = lookupResultQualifierCodeByCjsCode(bailQualifier)?.description
        if (description) {
          this.bailConditions.push(description)
        }
      })
    }

    return offence
  }

  execute(): OffencesResult {
    const spiOffences = this.courtResult.Session.Case.Defendant.Offence
    const offences = spiOffences.map(this.populateOffence).filter((offence) => !!offence) as Offence[]
    this.bailConditions = this.hearingDefendantBailConditions.concat(this.bailConditions)

    return { offences, bailConditions: this.bailConditions }
  }
}
