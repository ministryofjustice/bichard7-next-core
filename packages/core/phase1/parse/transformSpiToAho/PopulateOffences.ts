import type { Offence, OffenceCode } from "../../../types/AnnotatedHearingOutcome"
import { lookupAlcoholLevelMethodBySpiCode, lookupResultQualifierCodeByCjsCode } from "../../dataLookup"
import { COMMON_LAWS, INDICTMENT } from "../../lib/offenceTypes"
import resultCodeIsOnStopList from "../../lib/result/resultCodeIsOnStopList"
import type { OffenceParsedXml, ResultedCaseMessageParsedXml, SpiResult } from "../../types/SpiResult"
import PopulateOffenceResults from "./PopulateOffenceResults"
import removeSeconds from "./removeSeconds"

const enteredInErrorResultCode = 4583 // Hearing Removed
const dontKnowValue = "D"
const adjournmentSineDieResultCode = 2007
// TODO: Refactor into an enum
const timeRange = {
  ON_OR_IN: 1,
  BEFORE: 2,
  AFTER: 3,
  BETWEEN: 4,
  ON_OR_ABOUT: 5,
  ON_OR_BEFORE: 6
}

export interface OffencesResult {
  offences: Offence[]
  bailConditions: string[]
}

const adjournmentSineDieConditionMet = (spiResults: SpiResult[]) => {
  let a2007ResultFound = false
  let aFailConditionResultFound = false

  spiResults.forEach((result) => {
    const resultCode = result.ResultCode !== undefined ? Number(result.ResultCode) : undefined
    if (resultCode === adjournmentSineDieResultCode) {
      a2007ResultFound = true
    } else if (!resultCodeIsOnStopList(resultCode ?? 1000)) {
      aFailConditionResultFound = true
    }
  })

  return a2007ResultFound && !aFailConditionResultFound
}

const hasEnteredInErrorResult = (spiOffence: OffenceParsedXml): boolean =>
  spiOffence.Result.some(
    (result) => result.ResultCode !== undefined && Number(result.ResultCode) === enteredInErrorResultCode
  )

export default class {
  private bailConditions: string[] = []

  private isAdjournmentSineDieConditionMet = false

  constructor(private courtResult: ResultedCaseMessageParsedXml) {}

  private getOffenceReason = (spiOffenceCode: string): OffenceCode => {
    const spiOffenceCodeLength = spiOffenceCode.length
    const offenceCode: Pick<OffenceCode, "Reason" | "FullCode"> = {
      Reason: spiOffenceCodeLength > 4 ? spiOffenceCode.substring(4, Math.min(7, spiOffenceCodeLength)) : "",
      ...(spiOffenceCodeLength > 7 ? { Qualifier: spiOffenceCode.substring(7) } : {}),
      FullCode: spiOffenceCode
    }

    if (spiOffenceCode.startsWith(COMMON_LAWS)) {
      return {
        __type: "CommonLawOffenceCode",
        CommonLawOffence: COMMON_LAWS,
        ...offenceCode
      }
    }
    // Note: We do not receive indictment offences in magistrate courts but this is left for completeness
    if (spiOffenceCode.startsWith(INDICTMENT)) {
      return {
        __type: "IndictmentOffenceCode",
        Indictment: INDICTMENT,
        ...offenceCode
      }
    }

    return {
      __type: "NonMatchingOffenceCode",
      ActOrSource: spiOffenceCodeLength < 2 ? spiOffenceCode : spiOffenceCode.substring(0, 2),
      Year: spiOffenceCodeLength > 2 ? spiOffenceCode.substring(2, Math.min(4, spiOffenceCodeLength)) : "",
      ...offenceCode
    }
  }

  private populateOffence = (spiOffence: OffenceParsedXml): Offence => {
    const {
      BaseOffenceDetails: {
        OffenceCode: spiOffenceCode,
        OffenceSequenceNumber: spiOffenceSequenceNumber,
        ArrestDate: spiArrestDate,
        ChargeDate: spiChargeDate,
        LocationOfOffence: spiLocationOfOffence,
        OffenceWording: spiOffenceWording,
        AlcoholRelatedOffence: spiAlcoholRelatedOffence,
        OffenceTiming: { OffenceDateCode: spiOffenceDateCode, OffenceStart: spiOffenceStart, OffenceEnd: spiOffenceEnd }
      },
      ConvictionDate: spiConvictionDate,
      Result: spiResults
    } = spiOffence

    const { results, bailQualifiers } = new PopulateOffenceResults(this.courtResult, spiOffence).execute()

    if (bailQualifiers.length > 0) {
      bailQualifiers.forEach((bailQualifier) => {
        const description = lookupResultQualifierCodeByCjsCode(bailQualifier)?.description
        if (description) {
          this.bailConditions.push(description)
        }
      })
    }

    const offence: Offence = {
      CriminalProsecutionReference: {
        OffenceReason: {
          OffenceCode: this.getOffenceReason(spiOffenceCode ?? ""),
          __type: "NationalOffenceReason"
        }
      },
      ArrestDate: spiArrestDate ? new Date(spiArrestDate) : undefined,
      ChargeDate: spiChargeDate ? new Date(spiChargeDate) : undefined,
      ActualOffenceDateCode: spiOffenceDateCode?.toString(),
      ActualOffenceStartDate: {
        StartDate: new Date(spiOffenceStart.OffenceDateStartDate)
      },
      ActualOffenceEndDate: spiOffenceEnd ? { EndDate: new Date(spiOffenceEnd.OffenceEndDate) } : undefined,
      LocationOfOffence: spiLocationOfOffence,
      ActualOffenceWording: spiOffenceWording,
      CommittedOnBail: dontKnowValue,
      CourtOffenceSequenceNumber: Number(spiOffenceSequenceNumber),
      ConvictionDate: spiConvictionDate ? new Date(spiConvictionDate) : undefined,
      Result: results
    }

    if (spiAlcoholRelatedOffence) {
      offence.AlcoholLevel = {
        Amount: Number(spiAlcoholRelatedOffence.AlcoholLevelAmount),
        Method:
          lookupAlcoholLevelMethodBySpiCode(spiAlcoholRelatedOffence.AlcoholLevelMethod)?.cjsCode ??
          spiAlcoholRelatedOffence.AlcoholLevelMethod
      }
    }

    if (spiOffenceStart?.OffenceStartTime) {
      const spiOffenceStartTime = removeSeconds(spiOffenceStart.OffenceStartTime)

      const offenceDateCode = Number(spiOffenceDateCode)
      const { ON_OR_IN, BEFORE, AFTER, ON_OR_ABOUT, ON_OR_BEFORE, BETWEEN } = timeRange
      if ([ON_OR_IN, BEFORE, AFTER, ON_OR_ABOUT, ON_OR_BEFORE].includes(offenceDateCode)) {
        offence.OffenceTime = spiOffenceStartTime
      } else if (offenceDateCode === BETWEEN) {
        offence.StartTime = spiOffenceStartTime
      }
    }

    if (spiOffenceEnd?.OffenceEndTime) {
      offence.OffenceEndTime = removeSeconds(spiOffenceEnd.OffenceEndTime)
    }

    if (!spiConvictionDate) {
      this.isAdjournmentSineDieConditionMet ||= adjournmentSineDieConditionMet(spiResults)

      if (this.isAdjournmentSineDieConditionMet) {
        offence.ConvictionDate = new Date(this.courtResult.Session.CourtHearing.Hearing.DateOfHearing)
      }
    }

    return offence
  }

  execute(): OffencesResult {
    const spiOffences = this.courtResult.Session.Case.Defendant.Offence
    const offences = spiOffences.filter((o) => !hasEnteredInErrorResult(o)).map(this.populateOffence)

    return { offences, bailConditions: this.bailConditions }
  }
}
