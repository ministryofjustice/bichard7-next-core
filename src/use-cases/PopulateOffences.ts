import {
  ADJOURNMENT_SINE_DIE_RESULT_CODE_STRING,
  COMMON_LAWS,
  DONT_KNOW_VALUE,
  ENTERED_IN_ERROR_RESULT_CODE,
  INDICTMENT,
  STOP_LIST,
  TIME_RANGE
} from "src/lib/properties"
import type { Offence, OffenceCode } from "src/types/AnnotatedHearingOutcome"
import type { OffenceParsedXml, ResultedCaseMessageParsedXml } from "src/types/IncomingMessage"
import removeSeconds from "src/utils/removeSeconds"
import {
  lookupAlcoholLevelMethodBySpiCode,
  lookupOffenceCodeByCjsCode,
  lookupResultQualifierCodeByCjsCode
} from "./dataLookup"

import PopulateOffenceResults from "./PopulateOffenceResults"

export interface OffencesResult {
  offences: Offence[]
  bailConditions: string[]
}

export default class {
  private adjournmentSineDieConditionMet = false

  private bailConditions: string[] = []

  constructor(private courtResult: ResultedCaseMessageParsedXml, private hearingDefendantBailConditions: string[]) {}

  private getOffenceReason = (spiOffenceCode: string): OffenceCode => {
    const spiOffenceCodeLength = spiOffenceCode.length
    const offenceCode: Partial<OffenceCode> = {
      Reason: spiOffenceCodeLength > 4 ? spiOffenceCode.substring(4, Math.min(7, spiOffenceCodeLength)) : "",
      ...(spiOffenceCodeLength > 7 ? { Qualifier: spiOffenceCode.substring(7) } : {}),
      FullCode: spiOffenceCode
    }

    if (spiOffenceCode.startsWith(COMMON_LAWS)) {
      offenceCode.__type = "CommonLawOffenceCode"
      if (offenceCode.__type === "CommonLawOffenceCode") {
        offenceCode.CommonLawOffence = COMMON_LAWS
      }
    } else if (spiOffenceCode.startsWith(INDICTMENT)) {
      offenceCode.__type = "IndictmentOffenceCode"
      if (offenceCode.__type === "IndictmentOffenceCode") {
        offenceCode.Indictment = INDICTMENT
      }
    } else {
      offenceCode.__type = "NonMatchingOffenceCode"
      if (offenceCode.__type === "NonMatchingOffenceCode") {
        offenceCode.ActOrSource = spiOffenceCodeLength < 2 ? spiOffenceCode : spiOffenceCode.substring(0, 2)

        offenceCode.Year =
          spiOffenceCodeLength > 2 ? spiOffenceCode.substring(2, Math.min(4, spiOffenceCodeLength)) : ""
      }
    }

    return offenceCode as OffenceCode
  }

  private populateOffence = (spiOffence: OffenceParsedXml): Offence | undefined => {
    const offence = {} as Offence
    const {
      BaseOffenceDetails: {
        OffenceCode: spiOffenceCode,
        OffenceSequenceNumber: spiOffenceSequenceNumber,
        ArrestDate: spiArrestDate,
        ChargeDate: spiChargeDate,
        LocationOfOffence: spiLocationOfOffence,
        OffenceWording: spiOffenceWording,
        AlcoholRelatedOffence: spiAlcoholRelatedOffence,
        OffenceTiming: { OffenceDateCode: spiOffenceDateCode, OffenceStart: spiOffenceStart, OffenceEnd: spiOffenceEnd }
      },
      ConvictionDate: spiConvictionDate,
      Result: spiResults
    } = spiOffence

    spiResults.forEach((spiResult) => {
      if (spiResult.ResultCode === ENTERED_IN_ERROR_RESULT_CODE) {
        console.log(`Offence entered in error: ${spiOffenceCode}`)
        return undefined
      }
    })

    const OffenceCode = this.getOffenceReason(spiOffenceCode ?? "")
    offence.CriminalProsecutionReference = {
      OffenceReason: {
        OffenceCode,
        __type: "NationalOffenceReason"
      }
    }

    offence.ArrestDate = spiArrestDate ? new Date(spiArrestDate) : undefined
    offence.ChargeDate = spiChargeDate ? new Date(spiChargeDate) : undefined
    offence.ActualOffenceDateCode = spiOffenceDateCode?.toString()
    offence.ActualOffenceStartDate = {
      StartDate: new Date(spiOffenceStart.OffenceDateStartDate)
    }
    offence.ActualOffenceEndDate = {
      EndDate: spiOffenceEnd ? new Date(spiOffenceEnd.OffenceEndDate) : undefined
    }
    offence.LocationOfOffence = spiLocationOfOffence
    offence.ActualOffenceWording = spiOffenceWording

    if (spiAlcoholRelatedOffence) {
      offence.AlcoholLevel = {
        Amount: spiAlcoholRelatedOffence.AlcoholLevelAmount.toString(),
        Method:
          lookupAlcoholLevelMethodBySpiCode(spiAlcoholRelatedOffence.AlcoholLevelMethod)?.cjsCode ??
          spiAlcoholRelatedOffence.AlcoholLevelMethod
      }
    }

    if (spiOffenceStart?.OffenceStartTime) {
      const spiOffenceStartTime = removeSeconds(spiOffenceStart.OffenceStartTime)

      const { ON_OR_IN, BEFORE, AFTER, ON_OR_ABOUT, ON_OR_BEFORE, BETWEEN } = TIME_RANGE
      if ([ON_OR_IN, BEFORE, AFTER, ON_OR_ABOUT, ON_OR_BEFORE].includes(spiOffenceDateCode)) {
        offence.OffenceTime = spiOffenceStartTime
      } else if (spiOffenceDateCode === BETWEEN) {
        offence.StartTime = spiOffenceStartTime
      }
    }

    if (spiOffenceEnd?.OffenceEndTime) {
      offence.OffenceEndTime = removeSeconds(spiOffenceEnd.OffenceEndTime)
    }

    offence.ConvictionDate = spiConvictionDate ? new Date(spiConvictionDate) : undefined
    offence.CommittedOnBail = DONT_KNOW_VALUE
    offence.CourtOffenceSequenceNumber = spiOffenceSequenceNumber

    if (!spiConvictionDate) {
      const a2007ResultFound = spiResults.some(
        (spiResult) => spiResult.ResultCode?.toString() === ADJOURNMENT_SINE_DIE_RESULT_CODE_STRING
      )
      const aFailConditionResultFound = spiResults.some(
        (spiResult) => spiResult.ResultCode && STOP_LIST.includes(spiResult.ResultCode)
      )

      if (a2007ResultFound && !aFailConditionResultFound) {
        this.adjournmentSineDieConditionMet = true
      }

      if (this.adjournmentSineDieConditionMet) {
        offence.ConvictionDate = new Date(this.courtResult.Session.CourtHearing.Hearing.DateOfHearing)
      }
    }

    const { results, bailQualifiers } = new PopulateOffenceResults(this.courtResult, spiOffence).execute()
    offence.Result = results

    if (this.hearingDefendantBailConditions.length > 0 && bailQualifiers.length > 0) {
      bailQualifiers.forEach((bailQualifier) => {
        const description = lookupResultQualifierCodeByCjsCode(bailQualifier)?.description
        if (description) {
          this.bailConditions.push(description)
        }
      })
    }

    const offenceCode = lookupOffenceCodeByCjsCode(spiOffenceCode)
    if (offenceCode) {
      offence.RecordableOnPNCindicator = offenceCode.recordableOnPnc === "Y"
    }

    return offence
  }

  execute(): OffencesResult {
    const spiOffences = this.courtResult.Session.Case.Defendant.Offence
    const offences = spiOffences.map(this.populateOffence).filter((offence) => !!offence) as Offence[]
    this.bailConditions = this.hearingDefendantBailConditions.concat(this.bailConditions)

    return { offences, bailConditions: this.bailConditions }
  }
}
