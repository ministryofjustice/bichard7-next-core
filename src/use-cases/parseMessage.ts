import { XMLParser } from "fast-xml-parser"
import type { ResultedCaseMessageParsedXml } from "src/types/IncomingMessage"
import { incomingMessageParsedXmlSchema } from "src/types/IncomingMessage"

export default (message: string): ResultedCaseMessageParsedXml => {
  const options = {
    ignoreAttributes: false,
    removeNSPrefix: true
  }

  const parser = new XMLParser(options)
  const rawParsedObj = parser.parse(message) as unknown
  const parsedObj = incomingMessageParsedXmlSchema.parse(rawParsedObj)
  const rcm = parsedObj.DeliverRequest.Message.ResultedCaseMessage
  // if (!rcm.Session.Case.Defendant.Offence) {
  //   rcm.Session.Case.Defendant.Offence = []
  // }
  // if (!Array.isArray(rcm.Session.Case.Defendant.Offence)) {
  //   rcm.Session.Case.Defendant.Offence = [rcm.Session.Case.Defendant.Offence]
  // }
  // rcm.Session.Case.Defendant.Offence.forEach((offence) => {
  //   if (!offence.Result) {
  //     offence.Result = []
  //   }
  //   if (!Array.isArray(offence.Result)) {
  //     offence.Result = [offence.Result]
  //   }
  // })
  return rcm
}
