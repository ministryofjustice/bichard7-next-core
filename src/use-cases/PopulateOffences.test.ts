import { readFileSync } from "fs"
import parseMessage from "./parseMessage"
import PopulateOffences from "./PopulateOffences"

describe("populateDefendant", () => {
  const message = readFileSync("test-data/input-message-001.xml", "utf-8")
  const courtResult = parseMessage(message)

  it("should transform SPI Offences to Hearing Outcome Offences", () => {
    const result = new PopulateOffences(
      courtResult.Session.Case.Defendant.Offence,
      courtResult.Session.CourtHearing.Hearing.DateOfHearing
    ).execute()

    expect(result).toBeDefined()
    expect(result).toMatchSnapshot()
  })
})
