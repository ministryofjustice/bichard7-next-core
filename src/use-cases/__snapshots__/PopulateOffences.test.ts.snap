// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopulateOffences should transform SPI Offences to Hearing Outcome Offences 1`] = `
Object {
  "bailConditions": Array [],
  "offences": Array [
    Object {
      "ActualOffenceDateCode": "1",
      "ActualOffenceEndDate": Object {
        "EndDate": undefined,
      },
      "ActualOffenceStartDate": Object {
        "StartDate": "2010-11-28",
      },
      "ActualOffenceWording": "Attempt to rape a girl aged 13 / 14 / 15 / years of age - SOA 2003.",
      "ArrestDate": "2010-12-01",
      "ChargeDate": "2010-12-02",
      "CommittedOnBail": "D",
      "ConvictionDate": "2011-09-26",
      "CourtOffenceSequenceNumber": "1",
      "CriminalProsecutionReference": Object {
        "OffenceReason": Object {
          "OffenceCode": Object {
            "ActOrSource": "SX",
            "Qualifier": "A",
            "Reason": "001",
            "Year": "03",
          },
        },
      },
      "LocationOfOffence": "Kingston High Street",
      "Result": Array [
        Object {
          "CJSresultCode": "1000",
          "ConvictingCourt": undefined,
          "ModeOfTrialReason": "SUM",
          "NextHearingDate": undefined,
          "PleaStatus": "G",
          "ResultHearingDate": "2011-09-26",
          "ResultHearingType": "OTHER",
          "ResultQualifierVariable": Array [],
          "ResultVariableText": "Travel Restriction Order",
          "SourceOrganisation": Object {
            "BottomLevelCode": "01",
            "OrganisationUnitCode": "B01EF01",
            "SecondLevelCode": "01",
            "ThirdLevelCode": "EF",
            "TopLevelCode": "B",
          },
        },
      ],
    },
    Object {
      "ActualOffenceDateCode": "1",
      "ActualOffenceEndDate": Object {
        "EndDate": undefined,
      },
      "ActualOffenceStartDate": Object {
        "StartDate": "2010-11-28",
      },
      "ActualOffenceWording": "Rape of a Female",
      "ArrestDate": "2010-12-01",
      "ChargeDate": "2010-12-02",
      "CommittedOnBail": "D",
      "ConvictionDate": "2011-09-26",
      "CourtOffenceSequenceNumber": "2",
      "CriminalProsecutionReference": Object {
        "OffenceReason": Object {
          "OffenceCode": Object {
            "ActOrSource": "SX",
            "Reason": "001",
            "Year": "03",
          },
        },
      },
      "LocationOfOffence": "Kingston High Street",
      "Result": Array [
        Object {
          "CJSresultCode": "1000",
          "ConvictingCourt": undefined,
          "ModeOfTrialReason": "SUM",
          "NextHearingDate": undefined,
          "PleaStatus": "G",
          "ResultHearingDate": "2011-09-26",
          "ResultHearingType": "OTHER",
          "ResultQualifierVariable": Array [],
          "ResultVariableText": "defendant must never be allowed out",
          "SourceOrganisation": Object {
            "BottomLevelCode": "01",
            "OrganisationUnitCode": "B01EF01",
            "SecondLevelCode": "01",
            "ThirdLevelCode": "EF",
            "TopLevelCode": "B",
          },
        },
      ],
    },
    Object {
      "ActualOffenceDateCode": "1",
      "ActualOffenceEndDate": Object {
        "EndDate": undefined,
      },
      "ActualOffenceStartDate": Object {
        "StartDate": "2010-11-28",
      },
      "ActualOffenceWording": "Use a motor vehicle without third party insurance.",
      "ArrestDate": "2010-12-01",
      "ChargeDate": "2010-12-02",
      "CommittedOnBail": "D",
      "ConvictionDate": "2011-09-26",
      "CourtOffenceSequenceNumber": "3",
      "CriminalProsecutionReference": Object {
        "OffenceReason": Object {
          "OffenceCode": Object {
            "ActOrSource": "RT",
            "Reason": "191",
            "Year": "88",
          },
        },
      },
      "LocationOfOffence": "Kingston High Street",
      "Result": Array [
        Object {
          "CJSresultCode": "1000",
          "ConvictingCourt": undefined,
          "ModeOfTrialReason": "SUM",
          "NextHearingDate": undefined,
          "PleaStatus": "G",
          "ResultHearingDate": "2011-09-26",
          "ResultHearingType": "OTHER",
          "ResultQualifierVariable": Array [],
          "ResultVariableText": "Fined 100.",
          "SourceOrganisation": Object {
            "BottomLevelCode": "01",
            "OrganisationUnitCode": "B01EF01",
            "SecondLevelCode": "01",
            "ThirdLevelCode": "EF",
            "TopLevelCode": "B",
          },
        },
      ],
    },
  ],
}
`;
